app/validation.go

```diff
@@ -122,39 +203,62 @@ func checkParentsOfUndefined(p string, bosh propertyDefaults) bool {
 	return false
 }
 
-// collectManifestProperties returns a map merging the global and
-// per-instance-group properties/templates into a single structure.
-func collectManifestProperties(roleManifest *model.RoleManifest) map[string]string {
-	properties := make(map[string]string)
-
-	// Per-instance-group properties
-	for _, instanceGroup := range roleManifest.InstanceGroups {
-		for _, propertyDef := range instanceGroup.Configuration.Templates {
-			property := propertyDef.Key.(string)
-			template := propertyDef.Value
+func checkForSortedVariables(variables model.Variables) validation.ErrorList {
+	allErrs := validation.ErrorList{}
 
-			properties[property] = fmt.Sprintf("%v", template)
+	previousName := ""
+	for _, cv := range variables {
+		if cv.Name < previousName {
+			allErrs = append(allErrs, validation.Invalid("variables",
+				previousName,
+				fmt.Sprintf("Does not sort before '%s'", cv.Name)))
+		} else if cv.Name == previousName {
+			allErrs = append(allErrs, validation.Invalid("variables",
+				previousName, "Appears more than once"))
 		}
+		previousName = cv.Name
 	}
 
-	// And the global properties
-	for _, propertyDef := range roleManifest.Configuration.Templates {
-		property := propertyDef.Key.(string)
-		template := propertyDef.Value.(string)
+	return allErrs
+}
+
+// checkForUndefinedVariables checks that all configuration templates are
+// defined in the variables section
+func checkForUndefinedVariables(label string, templates map[string]model.ConfigurationTemplate, variableUsage map[string]int, isGlobal bool) validation.ErrorList {
```

Comment:
For this one, we may want it as a method of a `validator` struct that initializes and hols the state of `variableUsage`. Also, move the `range templates` to the caller, as this function don't need the whole thing (law of Demeter, or principle of least knowledge). Then we eliminate the need for passing `isGlobal` as well.

Reply:
Changed in 9b6136ede07c51872f28e995c24e15c78b4978df